File: Leetcode/1402 Reducing Dishes.ts
function maxSatisfaction(satisfactions: number[]): number {
    satisfactions.sort((a, b) => a - b);
    const n = satisfactions.length;
    const cache = new Array(n);
    for(let i = 0; i < n; i++) {
        cache[i] = new Array(n + 1).fill(Number.MIN_VALUE);
    }
    return maxSatisfactionDp(0, 1, satisfactions, cache);
}

// O(n^2) time DP solution
function maxSatisfactionDp(
    index: number, time: number, 
    satisfactions: number[], cache: number[][]
) {
    if (index == satisfactions.length) {
        return 0;
    }
    if (cache[index][time] != Number.MIN_VALUE) {
        return cache[index][time];
    }
    const ifCooked = satisfactions[index] * time + maxSatisfactionDp(
        index + 1, time + 1, satisfactions, cache
    );
    const ifNotCooked = maxSatisfactionDp(index + 1, time, satisfactions, cache);
    return cache[index][time] = Math.max(ifCooked, ifNotCooked);
}
                         
// O(n log n) time greedy solution
function maxSatisfactionGreedy(satisfactions: number[]) {
    satisfactions.sort((a, b) => a - b);
    let maxLikeTime = 0, suffixSum = 0;
    for (let i = satisfactions.length - 1; i >= 0; i--) {
        suffixSum += satisfactions[i];
        if(suffixSum < 0) {
            break;
        }
        maxLikeTime += suffixSum;
    }
    return maxLikeTime;
}
                         
                      

File: Leetcode/1402 Reducing Dishes.py
class Solution:
    def maxSatisfaction(self, satisfactions):
        n = len(satisfactions)
        cache = [[None for _ in range(n + 1)] for _ in range(n)]
        satisfactions.sort()
        return max_satisfaction_dp(0, 1, satisfactions, cache)

# O(n^2) time dp solution
def max_satisfaction_dp(index, time, satisfactions, cache):
    if index == len(satisfactions):
        return 0
    
    if cache[index][time] is not None:
        return cache[index][time]
    
    if_cooked = (
        time * satisfactions[index]
        + max_satisfaction_dp(index + 1, time + 1, satisfactions, cache)
    )
    
    if_not_cooked = max_satisfaction_dp(index + 1, time, satisfactions, cache)

    cache[index][time] = max(if_cooked, if_not_cooked)
    return cache[index][time]
  
  
  # O(n log n) time greedy solution
  def max_satisfaction_greedy(satisfactions):
    satisfactions.sort(reverse=True)
    prefix_sum, max_like_time  = 0, 0
    for satisfaction in satisfactions:
        prefix_sum += satisfaction
        if prefix_sum < 0:
            break
        max_like_time += prefix_sum
    return max_like_time
